### PR TITLE
Issue #3786: enable JOSM in Travis CI

### PR DESCRIPTION
**Pull request by don-vip**
Try to enable Ant-based JOSM build in Travis CI, as per #3786 comments

### CODE REVIEW

**Comment by don-vip** [comment]
@romani the JOSM target seems to work. Other errors seem unrelated to my changes.

**Comment by strkkk** [comment]
@don-vip usually all "no-error" things are in wercker, not travis

**Comment by don-vip** [comment]
@strkkk Did it change? I took what @romani started in #3783

**Comment by strkkk** [comment]
Looks strange, I see in travis.sh things like
`no-exception-test-hbase`,
`no-exception-test-spring-framework`
and others. They are included in travis.sh but not added in travis.yml
In general, wercker is used for it.

**Comment by romani** [comment]
> They are included in travis.sh but not added in travis.yml

Travis was first CI, almost all validations were in Travis originally.
moving JOSM to wercker is good idea to workaround Travis issue.

**Comment by romani** [comment]
Looks like we have no problem with Travis any more, it passed - https://travis-ci.org/github/checkstyle/checkstyle/jobs/677141507#L8468

@don-vip , please lets continue this PR. Please review other failures in CI, all of them need to be resolved. 
We need to discuss to avoid cat on report and .... 

@pbludov , can we use our new CI to host this validation ?

**Comment by romani** [comment]
@don-vip, please convert PR from draft to ready for review PR.

**Comment by romani** [comment]
@don-vip , please resolve CIs violations.

**Comment by pbludov** [comment]
> @pbludov , can we use our new CI to host this validation ?

I think we can. @don-vip please move the changes to .ci/semaphore.sh and .ci/semaphore.yaml

**Comment by romani** [comment]
it is better to not burn contributors with all nuances of our CI, @don-vip , please finish update for Travis to make it work sooner.
Semaphore could be another bag of problems, we can do this step later on.

**Comment by romani** [comment]
@don-vip , please look at CIs failures and please address them. We need green CIs to do merge.

**Comment by don-vip** [comment]
@romani unless I am missing something, the CI errors are not related to JOSM. I have absolutely no idea what needs to be fixed. I can commit the required changes if you guide me.

**Comment by romani** [comment]
@don-vip , please do:
1) https://travis-ci.org/github/checkstyle/checkstyle/jobs/677792589#L297
just apply patch to your branch

2) please do what is suggested at https://travis-ci.org/github/checkstyle/checkstyle/jobs/677792591#L635

all such violations are caused by your update, from your code that you contribute. We need to fix it.

**Comment by don-vip** [comment]
@romani thank you, I have applied your advice. Please don't forget none of this is "my" code, I copied what you wrote in #3783

**Comment by romani** [comment]
Please add "-e" as suggested at https://travis-ci.org/github/checkstyle/checkstyle/jobs/682304023#L635
Please do.

I  have so much PRs to review per day, I do no remember all context. I look at PR each  time almost from scratch.

File: .ci/travis/travis.sh [hunk 1]
```diff
@@ -461,6 +461,31 @@ no-error-pmd)
...
+  export PATH=$ANT_HOME/bin/:$PATH
+  echo "building checkstyle ..."
+  mvn clean package -Passembly
+  export CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.6.0:exec)
+  echo CS_version: ${CS_POM_VERSION}
+  echo "Checkouting sources ...."
+  svn -q export https://josm.openstreetmap.de/svn/trunk/ josm
+  cd josm
+  ant -v checkstyle
```
Comment (romani) [inline .ci/travis/travis.sh:478]: We need to use snapshot version of checkstyle in your project , checkstyle version is at CS_POM_VERSION

Comment (romani) [reply .ci/travis/travis.sh:479]: @don-vip , good , please wrap long line
see violation - https://travis-ci.org/github/checkstyle/checkstyle/jobs/682683301#L632

please run `mvn verify` before push, it would help to save a lot of time.

**Comment by romani** [comment]
now we have problem in your build item - https://travis-ci.org/github/checkstyle/checkstyle/jobs/682723821#L677 , please fix.

File: .ci/travis/travis.sh [hunk 2]
```diff
@@ -450,7 +450,7 @@ no-exception-test-alot-of-project1)
 
 no-error-pmd)
   CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
-                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.6.0:exec)
```
Comment (romani) [inline .ci/travis/travis.sh:453]: please remove this change (or move it to separate commit of the same PR), it is better to not do unrelated changes in same commit.

File: .ci/travis/travis.sh [hunk 3]
```diff
@@ -461,6 +461,33 @@ no-error-pmd)
...
+  echo CS_version: ${CS_POM_VERSION}
+  echo "Checkouting sources ...."
+  svn -q export https://josm.openstreetmap.de/svn/trunk/ josm
+  cd josm
+  sed -i -E "s/(name=\"checkstyle\" rev=\")([0-9]+\.[0-9]+(-SNAPSHOT)?)/\1${CS_POM_VERSION}/" \
+   tools/ivy.xml
+  ant -v checkstyle
+  ls -la
+  cat checkstyle-josm.xml
```
Comment (romani) [inline .ci/travis/travis.sh:482]: please  remove this `cat` to make logs huge, we do not need this.

```
$ git status
On branch don-vip/enable-josm
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   .ci/travis/travis.sh

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	apache-ant-1.10.7-bin.zip
	apache-ant-1.10.7/
	josm/

no changes added to commit (use "git add" and/or "git commit -a")
```

please clone/download all to `.ci-temp` folder, `git status`  should be clean.

Comment (romani) [reply .ci/travis/travis.sh:482]: @nmancus1 , please remind me where we stand on `git status` validation in CIs ?
do we have hanging PR for this ?

Comment (nmancus1) [reply .ci/travis/travis.sh:482]: No hanging PR, Shippable and Wercker are done, I just need to finish work on Travis, which I will complete this week.

**Comment by don-vip** [comment]
All is green now.

File: .ci/travis/travis.sh (hunk 3 above)
Comment (romani) [reply .ci/travis/travis.sh]: @don-vip ,

there is still a problem:
```
$ ./.ci/travis/travis.sh no-error-test-josm
....
$ git status
On branch don-vip/enable-josm
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	apache-ant-1.10.7-bin.zip
	apache-ant-1.10.7/
```
please download to `.ci-temp` folder. `git status` should be empty.

Comment (romani) [reply .ci/travis/travis.sh]: @nmancus1 , your issue is about cleanup of `.ci-temp`, it is not the same.
this problems is more critical, I moved `git status` validation to https://github.com/checkstyle/checkstyle/issues/8228 .

File: .ci/travis/travis.sh [hunk 4]
```diff
@@ -461,6 +461,31 @@ no-error-pmd)
...
 
+no-error-test-josm)
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.6.0:exec)
+  echo CS_version: ${CS_POM_VERSION}
+  mkdir -p .ci-temp
+  cd .ci-temp
+  ANT_VERSION=1.10.7
+  wget http://www-us.apache.org/dist//ant/binaries/apache-ant-${ANT_VERSION}-bin.zip
+  unzip -q apache-ant-${ANT_VERSION}-bin.zip
```
Comment (romani) [inline .ci/travis/travis.sh:472]: please add `-o` to avoid:
```
2020-05-09 17:20:06 (1.61 MB/s) - ‘apache-ant-1.10.7-bin.zip.1’ saved [10186279/10186279]

replace apache-ant-1.10.7/bin/ant? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace apache-ant-1.10.7/bin/antRun? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace apache-ant-1.10.7/bin/antRun.pl? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace apache-ant-1.10.7/bin/complete-ant-cmd.pl? [y]es, [n]o, [A]ll, [N]one, [r]ename: A
Apache Ant(TM) version 1.10.7 compiled on September 1 2019
....
```

File: .ci/travis/travis.sh [hunk 5]
```diff
@@ -461,6 +461,31 @@ no-error-pmd)
...
+  export ANT_HOME=$(pwd)/apache-ant-${ANT_VERSION}
+  export PATH=$ANT_HOME/bin/:$PATH
+  svn -q export https://josm.openstreetmap.de/svn/trunk/ josm
```
Comment (romani) [inline .ci/travis/travis.sh:475]: during execution run at:
```
2020-05-09 17:19:25 (1.72 MB/s) - ‘apache-ant-1.10.7-bin.zip’ saved [10186279/10186279]

svn: E155000: Destination directory exists; please remove the directory or use --force to overwrite
svn: E155000: 'josm' already exists 
```
please use `--force` option.

File: .ci/travis/travis.sh (hunk 4 above)
Comment (romani) [reply .ci/travis/travis.sh]: I still do not see '-o' option for unzip.
Please explain.

Comment (romani) [reply .ci/travis/travis.sh]: Please do not resolve review items, I will resolve them as ack that fix is properly done. Please just reply "done" in each review item.

**Comment by romani** [comment]
I run into problem on local:
```
[checkstyle] [WARN] .../.ci-temp/josm/src/org/openstreetmap/josm/actions/JumpToAction.java:15:8: Unused import - javax.swing.BorderFactory. [UnusedImports]
[checkstyle] To process the files took 19417 ms.
[checkstyle] Total execution took 19825 ms.

BUILD SUCCESSFUL
Total time: 35 seconds
Checkstyle Errors:
<error line="15" column="8" severity="warning" message="Unused import - javax.swing.BorderFactory." source="com.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck"/>
Size of output:1
```

@don-vip , please explain this, if checkstyle is not enforce on each commit .... it is better for us to use stable revision.

**Comment by don-vip** [comment]
It is not enforced, no. I changed the script to checkout the current tested version.

File: .travis.yml [hunk 6]
```diff
@@ -317,6 +317,13 @@ jobs:
           - USE_MAVEN_REPO="true"
           - CMD="./.ci/travis/travis.sh no-error-pmd"
 
+    # No error testing - josm
```
Comment (romani) [inline .travis.yml:320]: `No violation testing - josm (ant task)`

the same for DESC below.

because build is failed even if there is WARN violations.

File: .travis.yml [hunk 7]
```diff
@@ -317,6 +317,13 @@ jobs:
...
+    - jdk: openjdk11
+      env:
+          - DESC="no error test on josm"
+          - USE_MAVEN_REPO="true"
+          - CMD="./.ci/travis/travis.sh no-error-test-josm"
```
Comment (romani) [inline .travis.yml:325]: `no-violation-test-josm`

File: .ci/travis/travis.sh [hunk 8]
```diff
@@ -461,6 +461,34 @@ no-error-pmd)
...
+  mkdir -p .ci-temp
+  cd .ci-temp
+  ANT_VERSION=1.10.7
+  wget -q http://www-us.apache.org/dist//ant/binaries/apache-ant-${ANT_VERSION}-bin.zip -O tmp.zip
+  unzip -o -q tmp.zip
+  rm tmp.zip
+  export ANT_HOME=$(pwd)/apache-ant-${ANT_VERSION}
+  export PATH=$ANT_HOME/bin/:$PATH
+  TESTED=`wget -q -O - https://josm.openstreetmap.de/wiki/TestedVersion?format=txt`
+  echo "JOSM $TESTED"
```
Comment (romani) [inline .ci/travis/travis.sh:477]: `echo "JOSM revision: $TESTED"`

File: .travis.yml (hunk 6 above)
Comment (don-vip) [reply .travis.yml]: done

File: .travis.yml (hunk 7 above)
Comment (don-vip) [reply .travis.yml]: done

File: .ci/travis/travis.sh (hunk 8 above)
Comment (don-vip) [reply .ci/travis/travis.sh]: done

**Comment by don-vip** [comment]
> Even travis is green (I relaunched task and it is still green, same revision 16239) .... On my local I have stable failure of this build:
> @don-vip , do you know what SVN prop should we set in Travis to reproduce a problem ?

Are you running macOS? I only fixed this error in [r16391](https://josm.openstreetmap.de/changeset/16391/josm/) because I got it on Windows but it shouldn't occur on Linux. We can force an specific EOL using [--native-eol ARG](http://svnbook.red-bean.com/en/1.7/svn.ref.svn.html#svn.ref.svn.sw.native_eol).

Two questions for you:
- do you want me to hardcode a JOSM revision? I cannot guarantee all future tested versions of JOSM will be without any checkstyle violation
- do you want me to harcode an EOL when exporting source code from SVN?

File: .travis.yml (hunk 6 above)
Comment (romani) [reply .travis.yml]: DESC is not updated.

**Comment by romani** [comment]
> Are you running macOS?

```
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)"
.... 
$ svn --version
svn, version 1.9.3 (r1718519)
   compiled Jul 26 2019, 15:51:17 on x86_64-pc-linux-gnu
....
```

Travis  has also https://docs.travis-ci.com/user/reference/xenial#version-control `1.9.3` .

>  We can force an specific EOL using --native-eol ARG.

I am not sure what is wrong with my local SVN, may be some old global variable of SVN is make a difference.
Lets make use this option, it would help us to become consistent in behavior.

> do you want me to hardcode a JOSM revision?

yes, in other case there is no reason for us to use this project in CI. 
We will execute validation you josm code on each our commit, if any extra violation happens, it will force us to resolve issue. We even send patches to such projejcts with fixes for new violations, so update of checkstyle version in project become always safe. But if project maintainers do not want to cleanup code, we will drop project from CI usage.

> do you want me to harcode an EOL when exporting source code from SVN?

I tried:
```
✘-1 ~/java/github/romani/checkstyle [don-vip/enable-josm L|✚ 1] 
05:52 $ git diff
diff --git a/.ci/travis/travis.sh b/.ci/travis/travis.sh
index b4fd309..82a7caa 100755
--- a/.ci/travis/travis.sh
+++ b/.ci/travis/travis.sh
@@ -475,7 +475,7 @@ no-violation-test-josm)
   export PATH=$ANT_HOME/bin/:$PATH
   TESTED=`wget -q -O - https://josm.openstreetmap.de/wiki/TestedVersion?format=txt`
   echo "JOSM revision: $TESTED"
-  svn -q --force export https://josm.openstreetmap.de/svn/trunk/ -r $TESTED josm
+  svn -q --force  --native-eol LF export https://josm.openstreetmap.de/svn/trunk/ -r $TESTED josm
   cd josm
   sed -i -E "s/(name=\"checkstyle\" rev=\")([0-9]+\.[0-9]+(-SNAPSHOT)?)/\1${CS_POM_VERSION}/" \
    tools/ivy.xml
```

still a problem. Please advice.

File: .travis.yml (hunk 6 above)
Comment (don-vip) [reply .travis.yml]: done

**Comment by don-vip** [comment]
All comments addressed. I hope it's finally good.

**Comment by romani** [comment]
it works on my side:
```
./.ci/travis/travis.sh no-violation-test-josm
...
BUILD SUCCESSFUL
Total time: 20 seconds
Checkstyle Errors:
Size of output:0
```

File: .ci/travis/travis.sh [hunk 9]
```diff
@@ -461,6 +461,36 @@ no-error-pmd)
...
   ;;
 
+no-violation-test-josm)
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.6.0:exec)
+  echo CS_version: ${CS_POM_VERSION}
+  mkdir -p .ci-temp
+  cd .ci-temp
+  ANT_VERSION=1.10.7
+  wget -q http://www-us.apache.org/dist//ant/binaries/apache-ant-${ANT_VERSION}-bin.zip -O tmp.zip
```
Comment (rnveach) [inline .ci/travis/travis.sh:471]: Can travis not supply us with ant? https://docs.travis-ci.com/user/languages/java/#projects-using-ant
Requiring an install like this may conflict with local users and prevent them from running this piece of the script locally.

File: .ci/travis/travis.sh [hunk 10]
```diff
@@ -461,6 +461,36 @@ no-error-pmd)
...
+  unzip -o -q tmp.zip
+  rm tmp.zip
+  export ANT_HOME=$(pwd)/apache-ant-${ANT_VERSION}
+  export PATH=$ANT_HOME/bin/:$PATH
+  # Uncomment to test current tested version instead of hardcoded version
+  #TESTED=`wget -q -O - https://josm.openstreetmap.de/wiki/TestedVersion?format=txt`
+  TESTED=16391
```
Comment (rnveach) [inline .ci/travis/travis.sh:478]: Don't you need to make sure commented TESTED is after currently uncommented TESTED. If you uncomment the top one, the bottom one will just overwrite it again.

File: .ci/travis/travis.sh (hunk 9 above)
Comment (romani) [reply .ci/travis/travis.sh:471]: Travis use https://travis-ci.org/github/checkstyle/checkstyle/jobs/686213037#L128
```
ant version
Apache Ant(TM) version 1.9.6 compiled on July 20 2018
```

Comment (rnveach) [reply .ci/travis/travis.sh:471]: If ant is supplied in travis execution, why do we need other version? Does this project require a specific version of ant?

Comment (romani) [reply .ci/travis/travis.sh:471]: @don-vip, please reply.

Comment (don-vip) [reply .ci/travis/travis.sh:471]: Again, I didn't wrote this code, you did. This PR is working for several days, I spent a lot of time to help you to finish what you started but I don't have time anymore to work on never-ending feedback on cosmetic changes.

Comment (romani) [reply .ci/travis/travis.sh:471]: unfortunately I cannot remember all code, I have 1K PRs where all need assistance to merge it.
Quality of code is not disputable item in our project, we are not hosting of some code, code need to be finished and do not make problems to other, all visible points must be resolved. It is hard .... but it is definition of low threshold for new contributors and ease of maintainability.
